### PR TITLE
Do not `shuffle` in `LightningDataModule.from_datasets` for `IterableDataset`

### LINKED ISSUE #7051: Allow IterableDataset to be passed to LightningDataModule.from_datasets
## 🚀 Feature

Allow `IterableDataset` to be passed to `LightningDataModule.from_datasets`

### Motivation

Currently, we cannot pass `IterableDataset` instances to `LightningDataModule`'s `from_datasets` method.

This is because because `IterableDataset` cannot be used with `DataLoader` instances that are instantiated with `shuffle=True`. 

Right now, `DataLoader` instances created by `LightningDataModule` are hard-coded to set `shuffle=True`: 

https://github.com/PyTorchLightning/pytorch-lightning/blob/402a258705c10c8ad57bfdc16c39a8420b1425ee/pytorch_lightning/core/datamodule.py#L379

I think allowing users to pass a `shuffle` parameter to  `LightningDataModule.from_datasets` is a good idea.

### Additional context

I've made the code changes and am prepared to submit a pull request.

### PR DESCRIPTION
**Pull request by seongminp**
## What does this PR do?

Fixes #7051

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/datamodule.py [hunk 1]
```diff
@@ -373,10 +375,10 @@ def dataloader(ds, shuffle=False):
 
         def train_dataloader():
             if isinstance(train_dataset, Mapping):
-                return {key: dataloader(ds, shuffle=True) for key, ds in train_dataset.items()}
+                return {key: dataloader(ds, shuffle=shuffle) for key, ds in train_dataset.items()}
```
Comment (tchaton) [inline pytorch_lightning/core/datamodule.py:378]: Won't it be better to do: `shuffle=not isinstane(ds, IterableDataset)` directly ?
As you might have Dataset and IterableDataset in the list or dictionaries.

Comment (seongminp) [reply pytorch_lightning/core/datamodule.py:378]: That makes more sense. I'll push the change you mentioned.

File: pytorch_lightning/core/datamodule.py [hunk 2]
```diff
@@ -373,10 +373,13 @@ def dataloader(ds, shuffle=False):
 
         def train_dataloader():
             if isinstance(train_dataset, Mapping):
-                return {key: dataloader(ds, shuffle=True) for key, ds in train_dataset.items()}
+                return {
```
Comment (carmocca) [inline pytorch_lightning/core/datamodule.py:376]: nit: can you reuse this var?

```
should_shuffle = not isinstance(ds, IterableDataset)
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -98,6 +98,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Added `LightningModule.lr_schedulers()` for manual optimization  ([#6567](https://github.com/PyTorchLightning/pytorch-lightning/pull/6567))
 
+- Added `shuffle` parameter to `LightningModule.from_datasets()` ([#7503](https://github.com/PyTorchLightning/pytorch-lightning/pull/7503))
```
Comment (carmocca) [inline CHANGELOG.md:101]: 2 spaces

```suggestion

- Added `shuffle` parameter to `LightningModule.from_datasets()` ([#7503](https://github.com/PyTorchLightning/pytorch-lightning/pull/7503))
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/core/datamodule.py (hunk 2 above)
Comment (seongminp) [reply pytorch_lightning/core/datamodule.py:376]: Since `should_shuffle` has to be checked for every instance of `Dataset`, I think there are two options for code reuse:

**Option 1: Declaring a `should_shuffle` function.**
option1

**Option 2: Declaring a `should_shuffle` function.**
option2

Option 2 seems to be a major change from existing code behavior (introduces ways to override `shuffle`).

Option 1 seems to be personal preference, so right now I opted for no change. Any suggestions would be wonderful!

Comment (carmocca) [reply pytorch_lightning/core/datamodule.py:377]: I prefer the variable than the function in this case

Comment (seongminp) [reply pytorch_lightning/core/datamodule.py:377]: Should I commit with option 2? Or leave as-is?

File: pytorch_lightning/core/datamodule.py [hunk 4]
```diff
@@ -374,10 +374,13 @@ def dataloader(ds, shuffle=False):
...
+                    key: dataloader(ds, shuffle=not isinstance(ds, IterableDataset))
+                    for key, ds in train_dataset.items()
+                }
             if isinstance(train_dataset, Sequence):
-                return [dataloader(ds, shuffle=True) for ds in train_dataset]
-            return dataloader(train_dataset, shuffle=True)
+                return [dataloader(ds, shuffle=not isinstance(ds, IterableDataset)) for ds in train_dataset]
+            return dataloader(train_dataset, shuffle=not isinstance(train_dataset, IterableDataset))
```
Comment (awaelchli) [inline pytorch_lightning/core/datamodule.py:383]:
```suggestion
                return [dataloader(ds, shuffle=(not isinstance(ds, IterableDataset))) for ds in train_dataset]
            return dataloader(train_dataset, shuffle=(not isinstance(train_dataset, IterableDataset)))
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/core/datamodule.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/core/datamodule.py:377]: Made the change for you :)

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
